src/com/github/Kraken3/AFKPGC/LagScanner.java

```diff
@@ -22,6 +27,121 @@
 	private static Map<String, Map<Long, LagScanner.Result>> cache = 
 			new HashMap<String, Map<Long, LagScanner.Result>>();
 
+	private static Map<String, TreeMap<Long, HashSet<Long>>> worstChunksUnchecked = 
+			new HashMap<String, TreeMap<Long, HashSet<Long>>>();
+	private static Map<String, TreeMap<Long, Long>> worstChunksUncheckedIdx = 
+			new HashMap<String, TreeMap<Long, Long>>();
+	private static Map<String, TreeSet<Long>> worstChunksChecked = 
+			new HashMap<String, TreeSet<Long>>();
+
+	/**
+	 * Gets another chunk. Goes world by world from worst chunk to best in the world.
+	 */
+	public static Location getNextBadChunk(MutableLong ml, boolean mark) {
+		synchronized(worstChunksChecked) {
+			for (String world : worstChunksUnchecked.keySet()) {
+				TreeMap<Long, HashSet<Long>> worldBad = worstChunksUnchecked.get(world);
+				if (worldBad != null && !worldBad.isEmpty()) {
+					for (Long score : worldBad.descendingKeySet()) {
+						HashSet<Long> locations = worldBad.get(score);
+						if (locations != null && !locations.isEmpty()) {
+							Long location = locations.iterator().next();
+							long chunkX = location >> 32L;

```

Comment:
Check the index creation code; the X coord is stored in the "high order" bits via arithmetic shift 32 to the "left". To retrieve, I need to simply shift arithmetically 32 to the "right". Preserving sign is important; >>> is an unsigned shift and will likely corrupt - vs + (if I'm recollecting correctly). Then simply removing the high order bits by shifting back and subtracting should retrieve Z.

Double Check it on civtest for me, but again, worked great in limited direct testing.